File: LeetCode46.cpp
#include <vector>
#include<iostream>
using namespace std;

void backtracking(vector<int>& nums, int level, vector<vector<int>>& ans){
    if(level == nums.size() - 1){
        ans.push_back(nums);
        return ;
    }
    for (int i = level; i < nums.size(); i++){
        swap(nums[i], nums[level]);
        backtracking(nums, level + 1, ans);
        swap(nums[i], nums[level]);
    }
    
}

vector<vector<int>> permute(vector<int>& nums){
    vector<vector<int>> ans;
    backtracking(nums, 0, ans);
    return ans;
}

int main(){
    vector<vector<int>> grid;
    vector<int> child;
    int rows = 0, t = 0, n = 0;
    cout << "number of nums: ";
    cin >> rows;
    for (int i = 1; i <= rows; i++){
        cin >> t;
        child.push_back(t);
    }
    grid = permute(child);
    n = grid.size();
    cout << "[";
    for (int i = 0; i < n; i++){
        cout << "[";
        for (int k = 0; k < rows - 1; k++){
            cout << grid.at(i).at(k) << ",";
        }
        cout << grid.at(i).at(rows - 1) << "]";
    }
    cout << "]";
    return 0;
}
// number of nums: 3
// 1 2 3
// [[1,2,3][1,3,2][2,1,3][2,3,1][3,2,1][3,1,2]]

File: LeetCode46.java
import java.util.Scanner;
import java.util.List;
import java.util.Collections;
import java.util.ArrayList;

public class LeetCode46{
    public static void main(String[] args) {
        LeetCode46 solution = new LeetCode46();
        Scanner scanner = new Scanner(System.in);
        int [] grid;
        int rows = 0, columns = 0;
        System.out.print("number of nums: ");
        rows = scanner.nextInt();
        grid = new int[rows];
        for (int i = 1; i <= rows; i++){
            grid[i-1] = scanner.nextInt();
        }
        solution.permute(grid);
        scanner.close();
    }
    public List<List<Integer>> permute(int[] nums) {
        List<List<Integer>> res = new ArrayList<List<Integer>>();

        List<Integer> output = new ArrayList<Integer>();
        for (int num : nums) {
            output.add(num);
        }

        int n = nums.length;
        backtrack(n, output, res, 0);
        return res;
    }

    public void backtrack(int n, List<Integer> output, List<List<Integer>> res, int first) {
        // 所有数都填完了
        if (first == n) {
            res.add(new ArrayList<Integer>(output));
        }
        for (int i = first; i < n; i++) {
            // 动态维护数组
            Collections.swap(output, first, i);
            // 继续递归填下一个数
            backtrack(n, output, res, first + 1);
            // 撤销操作
            Collections.swap(output, first, i);
        }
    }
}

// number of nums: 3
// 1 2 3
// [[1,2,3][1,3,2][2,1,3][2,3,1][3,2,1][3,1,2]]